var/spack/repos/builtin/packages/adios/package.py

```diff
@@ -12,30 +36,106 @@ class Adios(Package):
     # module swap PrgEnv-intel PrgEnv-$COMP
     # module load cray-netcdf/4.3.3.1
     # module load cray-hdf5/1.8.14
     # module load python/2.7.10
-    depends_on('hdf5')
-    depends_on('mxml')
+
+    depends_on('autoconf', type='build')
+    depends_on('automake', type='build')
+    depends_on('libtool', type='build')
+    depends_on('python', type='build')
+
+    depends_on('mpi', when='+mpi')
+    # shipped within ADIOS 1.10.0+
+    depends_on('mxml@2.9:')
+    # optional transformations
+    depends_on('zlib', when='+zlib')
+    depends_on('szip', when='+szip')
+    # optional transports
+    depends_on('hdf5', when='+hdf5')
+    depends_on('netcdf', when='+netcdf')
+
+    def validate(self, spec):
+        """
+        Checks if incompatible variants have been activated at the same time
+        :param spec: spec of the package
+        :raises RuntimeError: in case of inconsistencies
+        """
+        if '+fortran' in spec and not self.compiler.fc:
+            msg = 'cannot build a fortran variant without a fortran compiler'
+            raise RuntimeError(msg)
 
     def install(self, spec, prefix):
-        configure_args = ["--prefix=%s" % prefix,
-                          "--with-mxml=%s" % spec['mxml'].prefix,
-                          "--with-hdf5=%s" % spec['hdf5'].prefix,
-                          "--with-netcdf=%s" % os.environ["NETCDF_DIR"],
-                          "--with-infiniband=no",
-                          "MPICC=cc", "MPICXX=CC", "MPIFC=ftn",
-                          "CPPFLAGS=-DMPICH_IGNORE_CXX_SEEK"]
+        self.validate(spec)
+        # Handle compilation after spec validation
+        extra_args = []
+
+        # required, otherwise building its python bindings on ADIOS will fail
+        extra_args.append("CFLAGS=-fPIC")
+
+        # always build external MXML, even in ADIOS 1.10.0+
+        extra_args.append('--with-mxml=%s' % spec['mxml'].prefix)
+
+        if '+shared' in spec:
+            extra_args.append('--enable-shared')
+
+        if '+mpi' in spec:
+            extra_args.append('--with-mpi')
+        if '+infiniband' in spec:
+            extra_args.append('--with-infiniband')
+        else:
+            extra_args.append('--with-infiniband=no')
+
+        if '+fortran' in spec:
+            extra_args.append('--enable-fortran')
+        else:
+            extra_args.append('--disable-fortran')
+
+        if '+zlib' in spec:
+            extra_args.append('--with-zlib=%s' % spec['zlib'].prefix)
+        if '+szip' in spec:
+            extra_args.append('--with-szip=%s' % spec['szip'].prefix)
+        if '+hdf5' in spec:
+            extra_args.append('--with-hdf5=%s' % spec['hdf5'].prefix)
+        if '+netcdf' in spec:
+            extra_args.append('--with-netcdf=%s' % os.environ["NETCDF_DIR"])
 
         if spec.satisfies('%gcc'):
-            configure_args.extend(["CC=gcc", "CXX=g++", "FC=gfortran"])
+            extra_args.extend(["CC=gcc", "CXX=g++"])
```

Comment:
normally you don't need to specify compilers. You can check `build.env` in `opt/..././../adios-.../.spack` to see which environment variables are set. If you need to set CXX here, one should use Spack compiler wrappers, i.e. `'CXX=%s' % spack_cpp` (not sure about the name, could be `spack_c++`).  Same below, FC,F77 will be set.